sys/net/application_layer/nanocoap/sock.c

```diff
@@ -102,41 +136,100 @@ ssize_t nanocoap_sock_request(nanocoap_sock_t *sock, coap_pkt_t *pkt, size_t len
                 continue;
             }
 
-            break;
+            switch (coap_get_type(pkt)) {
+            case COAP_TYPE_RST:
+                /* TODO: handle different? */
+                return -EBADMSG;
+            case COAP_TYPE_CON:
+                /* TODO: send ACK response */
+                /* fall-through */
+            case COAP_TYPE_NON:
+                if (pkt->payload_len == 0) {
+                    return 0;
+                }
+                /* fall-through */
+            case COAP_TYPE_ACK:
+                if (pkt->payload_len == 0 && response) {
+                    state = STATE_AWAIT_RESPONSE;
+                    /* TODO: adjust timeout */
```

Comment:
That's difficult to do. A server really only should reply with an empty ACK if waiting for the content of the reply would not be sensible (e.g. because the client is bound to assume a timeout and perform needless retransmissions). The standard only has this to say about what the timeout should be for the separate response:

> Note also that, while the CoAP protocol itself
      does not make any specific demands here, there is an expectation
      that the response will come within a time frame that is reasonable
      from an application point of view.

Reply:
So it's best to leave this as is?

Reply:
Likely one should, as the standard indicates, shift the responsibility to the application code. But it would be a pain in the ass to let all callers provide a timeout as parameter even if they do not expect a separate respone at all. But semantic-wise, I guess that would be the way to go.

One could also provide a compile time macro with sone sensible default and let the application adjust it as needed. But anyone who has a use case where different timeouts for different requests are needed will start throwing rotten tomatos at us.